ContinousBenchmarking.md

```diff
@@ -0,0 +1,88 @@
+Continuous Benchmarking System
+======
+Kevin Cheng
+Initial Proposal 11/6/18
+
+
+# Background:
+## Context
+Benchmarking an algorithm is a compute intensive and time consuming chore that is essential for guaranteeing performance. 
+An algorithm can unknowingly drift towards an undesirable state if the laborious task of benchmarking is neglected for
+too long.  
+
+## Problem Statement
+We want to create a continous benchmarking system to automate this task.  
+
+## Scope
+Problem starts with a code changes to Garage. The system needs to run benchmarking at least nightly, generate a results page, and deploy that page.  Problem ends with a pubically viewable results page.
+
+
+## Goals and Non-Goals
+*This version of the proposal will focus on a minimally viable system. These goals will evolve once the initial goal is accomplished and we decide to bring in more complexity.
+#### Goals:
+- Update benchmarking scripts so they can be started in a consistent format and run to completion without manual intervention.
+- Focus on getting system up and running on a single machine.
+- Design a simple system for deploying the results page.
+- Allow garage developers to detect regressions in algorithm performance caused by code changes
+
+#### Non-Goals: 
+- Worry about scalability too early.  MVP is top priority.
+- Design for frugality over simplicity
+- Manage web servers
+
+## Design Overview
+*The input and output interfaces of the benchmarking scripts must be normalized (start commands, and results).
+This initial design will use a cron job script ran on a single local machine. On a nightly basis the script shall:
+1. pull the latest garage code. 
+2. Run commands to launch benchmarking jobs
+3. Detect when the jobs are completed
+4. Run script to generate results page with plots.
+5. Deploy results.html to Github pages or similar.
```

Comment:
Will we be merging to master and then building to verify nightly? (I would hope we test before merging, but wanted explicit clarity on this)

Plan of Action:
if tests pass, then merge to master?
if tests fail, then ... ?
if pull request PR1 passes, and we merge. And then PR2 is queued up next, will you rebase off of PR1 and then run test? Or run PR2 tests based off of master?

Reply:
There are standard procedures for these. Also, I feel this may be more fleshed out in the 'Detailed Design' section. @ryanjulian what level of detail are we expecting here?

Reply:
@gautams3  I think this system will not be connected to TravisCI in this first version.  I suggest we not gate any merges until the system has been running for sometime and is deemed to be reliable.

Reply:
At that point, we may want to decide how many hours of run time is acceptable as a gate.  ~24 hours is quite a lot, and is starting to stretch the definition of "continuous"

Reply:
I would be happy with ~7 days maximum